trieste/ask_tell_optimization.py

```diff
@@ -0,0 +1,181 @@
+
+        if isinstance(datasets, Dataset):
+            datasets = {OBJECTIVE: datasets}
+            # ignore below is due to MyPy not being able to handle overloads properly
+            model_specs = {OBJECTIVE: model_specs}  # type: ignore
+
+        # reassure the type checker that everything is tagged
+        datasets = cast(Dict[str, Dataset], datasets)
+        model_specs = cast(Dict[str, ModelSpec], model_specs)
+
+        if datasets.keys() != model_specs.keys():
+            raise ValueError(
+                f"datasets and model_specs should contain the same keys. Got {datasets.keys()} and"
+                f" {model_specs.keys()} respectively."
+            )
+
+        self._datasets = datasets
+        self._models = map_values(create_model, model_specs)
+
+        if fit_model:
+            # there is no default value for `fit_model` parameter
+            # to force the user to think about it
+            # before invoking ask/tell API
+            # there is no sensible default here:
+            # `False` may result in suboptimal performance
+            # `True` may result in two model training calls per optimization step
+            for tag, model in self._models.items():
+                dataset = datasets[tag]
+                model.update(dataset)
+                model.optimize(dataset)
+
+    @classmethod
+    def from_record(
+        cls, search_space: SP, acquisition_rule: AcquisitionRule[S, SP], record: Record[S]
+    ) -> AskTellOptimizer[S, SP]:
+        """Creates new :class:`~AskTellOptimizer` instance from provided optimization state.
+
+        :param search_space: The space over which to search.
+        :param acquisition_rule: The acquisition rule, which defines how to search for a new point
+            on each optimization step.
+        :param record: Optimization state record.
+        """
+        # here we are recovering previously saved optimization state
+        # there the model was already trained
+        # thus there is no need to train it again
+        fit_model = False
+        return cls(
+            search_space,
+            record.datasets,
+            record.models,
+            fit_model,
+            acquisition_rule,
+            record.acquisition_state,
+        )
+
+    def ask(self) -> TensorType:
+        """Suggests a point (or points in batch mode) to observe by optimizing the acquisition function.
+        If the acquisition is stateful, its state is saved.
+
+        :return: A :class:`TensorType` instance representing suggested point(s).
+        """
+        points_or_stateful = self._acquisition_rule.acquire(
+            self._search_space, self._datasets, self._models
+        )
+
+        if callable(points_or_stateful):
+            self._acquisition_state, query_points = points_or_stateful(self._acquisition_state)
+        else:
+            query_points = points_or_stateful
+
+        return query_points
```

Comment:
Q: Do we want to protect against calling ask twice in a row without an intermediate tell? Especially with stateful rules? Not sure.

Reply:
i think we actually might want several asks in a row sometimes. i'll next work on async batch BO, where this is likely to be the case